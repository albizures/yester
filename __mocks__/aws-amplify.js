
export const Auth = {
  federatedSignIn: jest.fn().mockResolvedValue({ sessionToken: '1234567' }),
  currentSession: jest.fn().mockResolvedValue({
    accessToken: {
      jwtToken: '12345qwert',
    },
  }),
  signIn: jest.fn().mockResolvedValue({}),
}

export default {
  configure: jest.fn(),
}
