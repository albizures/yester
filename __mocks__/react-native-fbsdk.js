export const LoginManager = {
  logInWithReadPermissions: jest.fn().mockResolvedValue({ isCancelled: false }),
}

export const AccessToken = {
  getCurrentAccessToken: jest.fn().mockResolvedValue({
    accessToken: '2345678', expirationTime: 12345678,
  }),
}

export const GraphRequest = jest.fn().mockImplementation((route, config, callback) => {
  callback(undefined, {})
})

export const mockStart = jest.fn()
export const mockAddRequest = jest.fn().mockImplementation(() => ({
  start: mockStart,
}))
export const GraphRequestManager = jest.fn().mockImplementation(() => ({
  addRequest: mockAddRequest,
}))
