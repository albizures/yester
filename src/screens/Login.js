import { Auth } from 'aws-amplify'
import PropTypes from 'prop-types'
import React, { Component } from 'react'
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view'
import { Alert, View, StyleSheet, Dimensions, Platform } from 'react-native'
import icons from '../utils/icons'
import colors from '../utils/colors'
import { logIn, postAPIUser } from '../utils/session'
import Button from '../components/Button'
import { Heading2, Heading3, Description } from '../components'
import Container from '../components/Container'
import withFBLogin from '../components/withFBLogin'
import TextDivider from '../components/TextDivider'
import TextInput from '../components/TextInput'
import TopBar from '../components/TopBar'
import { translate } from '../components/Translate'
import { screen } from '../utils/analytics'
import DeviceInfo from 'react-native-device-info'
import debugFactory from 'debug'

const debugInfo = debugFactory('yester:Login:info')
const debugError = debugFactory('yester:Login:error')

class Login extends Component {
  static propTypes = {
    navigation: PropTypes.object.isRequired,
    onLoginWithFB: PropTypes.func.isRequired,
  }

  state = {
    email: '',
    password: '',
  }

  componentDidMount () {
    screen('Login', {})
  }

  onFBNativeLogin = async () => {
    const { onLoginWithFB, navigation } = this.props

    try {
      const { token, expires, profile } = await onLoginWithFB()
      debugInfo('Profile: ', profile, expires)
      await Auth.federatedSignIn('facebook', { token, expires_at: expires }, profile)

      const currentUser = await Auth.currentAuthenticatedUser()
      await postAPIUser({
        email: currentUser.email,
        given_name: profile['first_name'],
        family_name: profile['last_name'],
        platform: Platform.OS,
        build: DeviceInfo.getBuildNumber(),
        version: DeviceInfo.getVersion(),
        email_verified: true,
      })

      return navigation.navigate('AppLoading')
    } catch (error) {
      debugError('NativeFB Login', error)
      Alert.alert(translate('login.error.facebook'))
    }
  }

  onFBWebView = async () => {
    this.props.navigation.navigate('FBWebView')
  }

  onLogin = async () => {
    const { navigation } = this.props
    const { email, password } = this.state
    try {
      await logIn(email, password)

      // When isn't federatedSignIn, the user comes from CognitoUP
      // so it has got to use 'attributes' in order to get email.
      const currentUser = await Auth.currentAuthenticatedUser()
      await postAPIUser({
        email: currentUser.attributes.email,
        platform: Platform.OS,
        build: DeviceInfo.getBuildNumber(),
        version: DeviceInfo.getVersion(),
      })

      return navigation.navigate('AppLoading')
    } catch (error) {
      debugError('Login', error)
      if (error.code === 'NotAuthorizedException') {
        Alert.alert(translate('login.error.notAuthorized'))
      } else if (error.code === 'UserNotFoundException') {
        Alert.alert(translate('login.error.userNotFound'))
      } else {
        Alert.alert(translate('login.error'))
      }
    }
  }

  onChange = (value, name) => {
    this.setState({
      [name]: value,
    })
  }

  onForgotPassword = () => {
    const { navigation } = this.props
    navigation.navigate('ForgotPassword')
  }

  onBack = () => {
    const { navigation } = this.props
    navigation.navigate('CreateAccount')
  }

  render () {
    const { email, password } = this.state
    const topBar = <TopBar title='createAccount.login' onBack={this.onBack} />
    return (
      <Container topBar={topBar}>
        <KeyboardAwareScrollView extraScrollHeight={170} enableOnAndroid>
          <View style={styles.view}>
            <View style={styles.topFlex}>
              <Button
                title='createAccount.continue'
                icon={icons.fb}
                onPress={this.onFBNativeLogin}
              />
              <Description
                keyName='createAccount.recommendation'
                style={styles.recommendationText}
              />
              <TextDivider>
                <Heading3 keyName='createAccount.or' />
              </TextDivider>
              <Heading2 keyName='login.loginAccount' style={styles.loginAccountText} />
            </View>

            <View style={styles.bottomFlex}>
              <TextInput
                title='signup.email'
                autoCapitalize='none'
                keyboardType='email-address'
                value={email}
                onChangeText={(text) => this.onChange(text, 'email')}
              />
              <TextInput
                title='signup.password'
                password
                value={password}
                onChangeText={(text) => this.onChange(text, 'password')}
              />
              <Description
                keyName='login.forgotPassword'
                onPress={this.onForgotPassword}
                style={styles.forgotPasswordText}
              />
              <Button onPress={this.onLogin} title='createAccount.login' type={Button.OUTLINED} />
            </View>
          </View>
        </KeyboardAwareScrollView>
      </Container>
    )
  }
}

const { height, width } = Dimensions.get('window')
const styles = StyleSheet.create({
  view: {
    width,
    height: height - 60,
    backgroundColor: colors.athensGray,
    paddingHorizontal: width * 0.07,
    paddingTop: height * 0.05,
    paddingBottom: height * 0.1,
  },
  topFlex: {
    flex: 4,
    alignItems: 'center',
    paddingBottom: height * 0.05,
  },
  bottomFlex: {
    flex: 6,
    alignItems: 'center',
  },
  recommendationText: {
    alignSelf: 'center',
    marginTop: height * 0.03,
    marginBottom: height * 0.045,
  },
  loginAccountText: {
    color: colors.governorBay,
    marginTop: height * 0.03,
  },
  forgotPasswordText: {
    fontWeight: 'bold',
    color: colors.governorBay,
    alignSelf: 'flex-start',
    paddingLeft: 10,
    marginBottom: height * 0.06,
  },
})

export default withFBLogin(Login)
