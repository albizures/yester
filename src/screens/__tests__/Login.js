import React from 'react';
import { shallow } from 'enzyme';
import { Auth } from 'aws-amplify';
import { Button, Alert } from 'react-native';
import renderer from 'react-test-renderer';
import { LoginManager } from 'react-native-fbsdk';

import Login from '../Login';

Alert.alert = jest.fn();

const navigation = { navigate: jest.fn() };

beforeEach(() => {
	Alert.alert.mockClear();
});

describe.skip('src/screens/Login.js', () => {
	it('should render', () => {
		const tree = renderer.create(<Login navigation={navigation} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	describe('when facebook button is pressed', () => {
		it('should login to cognito', async () => {
			const shallowLogin = shallow(<Login navigation={navigation} />);
			const shallowTouchableHighlights = shallowLogin.find(Button);
			const shallowButtonToHome = shallowTouchableHighlights.at(0);

			await shallowButtonToHome.prop('onPress')();

			expect(LoginManager.logInWithReadPermissions).toHaveBeenCalledTimes(1);
			expect(LoginManager.logInWithReadPermissions).toHaveBeenCalledWith([
				'email',
				'public_profile',
			]);
			expect(Auth.federatedSignIn).toHaveBeenCalledTimes(1);
			expect(Auth.federatedSignIn).toHaveBeenCalledWith(
				'facebook',
				expect.objectContaining({
					token: expect.any(String),
					expires_at: expect.any(Number),
				}),
				expect.any(Object),
			);
		});

		describe('when something fails', () => {
			it('should show an alert', async () => {
				const shallowLogin = shallow(<Login navigation={navigation} />);
				const shallowTouchableHighlights = shallowLogin.find(Button);
				const shallowButtonToHome = shallowTouchableHighlights.at(0);

				LoginManager.logInWithReadPermissions.mockRejectedValueOnce(
					new Error('some error'),
				);

				await shallowButtonToHome.prop('onPress')();
				expect(Alert.alert).toHaveBeenCalledTimes(1);
				expect(Alert.alert).toHaveBeenCalledWith('some error');
			});
		});

		describe('when the login is cancelled', () => {
			it('should show an alert', async () => {
				const shallowLogin = shallow(<Login navigation={navigation} />);
				const shallowTouchableHighlights = shallowLogin.find(Button);
				const shallowButtonToHome = shallowTouchableHighlights.at(0);

				LoginManager.logInWithReadPermissions.mockResolvedValue({
					isCancelled: true,
				});

				await shallowButtonToHome.prop('onPress')();
				expect(Alert.alert).toHaveBeenCalledTimes(1);
				expect(Alert.alert).toHaveBeenCalledWith('Login cancelled');
			});
		});
	});
});
