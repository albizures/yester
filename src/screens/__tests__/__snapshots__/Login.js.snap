// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/screens/Login.js should render 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
        "paddingTop": 20,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Array [
        Object {
          "marginTop": 40,
          "textAlign": "center",
        },
      ]
    }
  >
    LOG IN
  </Text>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        undefined,
        Object {
          "borderRadius": 24,
          "borderWidth": 2,
          "height": 50,
          "justifyContent": "center",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
        },
        Object {
          "backgroundColor": "#4241B3",
          "borderColor": "#4241B3",
        },
        Object {
          "width": 300,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/fb-no-oficial.png",
          }
        }
        style={
          Array [
            Object {
              "height": 20,
              "marginRight": 9,
              "width": 20,
            },
            Object {
              "tintColor": "#fff",
            },
          ]
        }
      />
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            Object {
              "color": "#fff",
            },
          ]
        }
      >
        Continue with Facebook
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
        "marginHorizontal": 10,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#D5D3DD",
            "borderColor": "#D5D3DD",
            "borderWidth": 1,
            "height": 0,
            "marginHorizontal": 11,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    />
    <View
      styles={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#313131",
              "fontSize": 20,
              "textAlign": "left",
            },
            undefined,
          ]
        }
      >
        Or
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#D5D3DD",
            "borderColor": "#D5D3DD",
            "borderWidth": 1,
            "height": 0,
            "marginHorizontal": 11,
          },
          Object {
            "flex": 1,
          },
        ]
      }
    />
  </View>
  <View
    style={
      Object {
        "borderColor": "#000",
        "borderStyle": "solid",
        "borderWidth": 1,
        "marginBottom": 10,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    style={
      Object {
        "borderColor": "#000",
        "borderStyle": "solid",
        "borderWidth": 1,
        "marginBottom": 10,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      secureTextEntry={true}
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        undefined,
        Object {
          "borderRadius": 24,
          "borderWidth": 2,
          "height": 50,
          "justifyContent": "center",
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 3,
            "width": 0,
          },
        },
        Object {
          "backgroundColor": "#fff",
          "borderColor": "#4241B3",
        },
        Object {
          "width": 300,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "fontSize": 18,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            Object {
              "color": "#4241B3",
            },
          ]
        }
      >
        Log In
      </Text>
    </View>
  </View>
</View>
`;
